### ISSUE TITLE
sketch_rnn: cost goes negative, no improvement in KL

### ISSUE THREAD

**Issue by silky** [tensorflow/magenta#945]
i'm playing around with a custom dataset in sketch rnn

i've set it up in (basically) the stroke-3 format; expect in my case i'm not using dx dy, but just simply x, y

i'm keeping the pen-state as a single boolean

however, before progress very far my network cost goes negative and after that no improvement in KL is seen:

![image](https://user-images.githubusercontent.com/129525/32704079-f602b61e-c853-11e7-927b-6f559415585e.png)

these are the hparams:

```
...
INFO:tensorflow:sketch-rnn
INFO:tensorflow:Hyperparams:                                                                                                                                                                                                                                                                
INFO:tensorflow:grad_clip = 1.0                                                                                                                                                                                                                                                             
INFO:tensorflow:conditional = True                                                                                                                                                                                                                                                          
INFO:tensorflow:min_learning_rate = 1e-05                                                                                                                                                                                                                                                   
INFO:tensorflow:num_mixture = 20                                                                                                                                                                                                                                                            
INFO:tensorflow:is_training = True                                                                                                                                                                                                                                                          
INFO:tensorflow:input_dropout_prob = 0.9                                                                                                                                                                                                                                                    
INFO:tensorflow:kl_decay_rate = 0.99995                                                                                                                                                                                                                                                     
INFO:tensorflow:kl_tolerance = 0.2                                                                                                                                                                                                                                                          
INFO:tensorflow:random_scale_factor = 0.15                                                                                                                                                                                                                                                  
INFO:tensorflow:decay_rate = 0.9999                                                                                                                                                                                                                                                         
INFO:tensorflow:use_recurrent_dropout = True                                                                                                                                                                                                                                                
INFO:tensorflow:num_steps = 10000000                                                                                                                                                                                                                                                        
INFO:tensorflow:use_output_dropout = False                                                                                                                                                                                                                                                  
INFO:tensorflow:max_seq_len = 250                                                                                                                                                                                                                                                           
INFO:tensorflow:z_size = 128                                                                                                                                                                                                                                                                
INFO:tensorflow:augment_stroke_prob = 0.0                                                                                                                                                                                                                                                   
INFO:tensorflow:learning_rate = 0.001                                                                                                                                                                                                                                                       
INFO:tensorflow:batch_size = 100                                                                                                                                                                                                                                                            
INFO:tensorflow:enc_model = layer_norm                                                                                                                                                                                                                                                      
INFO:tensorflow:use_input_dropout = False                                                                                                                                                                                                                                                   
INFO:tensorflow:dec_model = hyper                                                                                                                                                                                                                                                           
INFO:tensorflow:enc_rnn_size = 256                                                                                                                                                                                                                                                          
INFO:tensorflow:output_dropout_prob = 0.9                                                                                                                                                                                                                                                   
INFO:tensorflow:save_every = 500                                                                                                                                                                                                                                                            
INFO:tensorflow:kl_weight = 0.5                                                                                                                                                                                                                                                             
INFO:tensorflow:data_set = ['choreo-alt-data.npz']                                                                                                                                                                                                                                          
INFO:tensorflow:kl_weight_start = 0.01                                                                                                                                                                                                                                                      
INFO:tensorflow:dec_rnn_size = 2048                                                                                                                                                                                                                                                         
INFO:tensorflow:recurrent_dropout_prob = 0.9                                                                                                                                                                                                                                                
INFO:tensorflow:Loading data files.                                                                                                                                                                                                                                                         
INFO:tensorflow:Loaded 172490/5505/5505 from choreo-alt-data.npz                                                                                                                                                                                                                            
INFO:tensorflow:Dataset combined: 183500 (172490/5505/5505), avg len 234                                                                                                                                                                                                                    
INFO:tensorflow:model_params.max_seq_len 234.                                                                                                                                                                                                                                               
total images <= max_seq_len is 172490                                                                                                                                                                                                                                                       
total images <= max_seq_len is 5505                                                                                                                                                                                                                                                         
total images <= max_seq_len is 5505                                                                                                                                                                                                                                                         
INFO:tensorflow:normalizing_scale_factor 0.1818.
INFO:tensorflow:Model using gpu.
INFO:tensorflow:Input dropout mode = False.
INFO:tensorflow:Output dropout mode = False.
INFO:tensorflow:Recurrent dropout mode = True.
INFO:tensorflow:Model using gpu.
INFO:tensorflow:Input dropout mode = 0.
INFO:tensorflow:Output dropout mode = 0.
INFO:tensorflow:Recurrent dropout mode = 0.
```

any thoughts?

**Comment by hardmaru** [comment]
Hi Noon van der Silk,

There is going to be a trade off between the reconstruction loss and the KL
loss.

What numbers did you ultimately get for the reconstruction loss and KL
loss? And can you use the trained model to generate meaningful sketches?
How large is your dataset?

If you want to have a better KL loss, sooner in training, I would recommend
try setting kl_decay_rate to 0.9990. In addition, setting kl_weight to 1.0
will lower the KL loss.

**Comment by odie2630463** [comment]
@hardmaru why reconstruction loss will be negative? I think it just a normal distribution pdf , how it will be greater than one , and make reconstruction loss become negative? Did I something wrong? Thk

**Comment by hardmaru** [comment]
Hi!

In the paper, the loss function is a combination of gaussian and discrete pdf (for pen strokes, and end-of-drawing etc), so it can be negative.

Thks
